Support passive materializaion case1

navyxliu opened the pull request:
This patch materializes an object for its predecessors induced by the incoming edge.  The challenging part is that C2 Parse doesn't retain the compile-time states for predecessors. we have to split it from current map on our own. We introduce a shim region and generates a stationary map (SafePointNode*) for it. 

In Region::ideal(), we have to normalize the region node. A CFG node can't control over multiple inputs of a region node. we have to compact it. Essentially, shim region is the common code block for them. 
![passive_materialize_casea](https://github.com/navyxliu/jdk/assets/2386768/168d8b9a-9e21-4984-a017-7a5a46460b25)

Commit:
wip

Commit:
Refactor the interface of passive materialization

Commit:
Support 2 cases for shim construction

Commit:
Don't do anything in merge_at_phi_creation

Commit:
remove trim

Commit:
Update equals().

Commit:
fix bugs in implementation.

Commit:
Remove PEAState::validate().

it has not been functional.

Commit:
don't update newin in merge_at_phi_creation.

Commit:
lock down phi nodes after merging.

Commit:
enable passive_materialization_for_predecessor in merge_at_phi_creation.

Commit:
Normalize region node.

this patch fixes build_cfg() error.
in RegionNode::ideal(), we delete the redundant edges. they are
introduced by passive materialization.

n
| \(del)
---------
|region |
---------
         \  |  /(del)
         ------
         | phi|
         ------

Commit:
JVM-2523: bad graph detected caused by passive materialization.

in Parse::return_current(), we should merge the return values for
_exits.

Commit:
update for  PredictedCallGenerator::generate().

in passive_materialization, we support lock down node.
our process is just like Compile::gvn_replace_by().

one special case is MergeNode. we don't merge it until we complete all
other nodes. this is same as merge_common().

Commit:
JVM-2524: Crashed at java.util.concurrent.CompletableFuture::andTree

Commit:
Remove run_merge_if_else_paranoid.sh from problemlist.

caojoshua commented on src/hotspot/share/opto/c2_globals.hpp:
```diff
@@ -489,6 +489,9 @@
           "Perform partial escape analysis")                                \
                                                                             \
+  product(bool, EnablePassiveMaterialization, true, EXPERIMENTAL,           \
+          "Do passive materialization")                                     \
```
I'm in favor of just always doing passive materialization. Without it PEA has regressions.

caojoshua commented on src/hotspot/share/opto/cfgnode.cpp:
```diff
@@ -550,6 +579,40 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
           }
         }
       }
+    } else {
```
Can you add a comment for this block? Not sure its purpose.

caojoshua commented on src/hotspot/share/opto/cfgnode.hpp:
```diff
@@ -83,6 +83,7 @@ class RegionNode : public Node {
 
 private:
   bool _is_unreachable_region;
+  bool _shim; // inserted by PEA
```
Maybe we should write out `Partial Escape Analysis`. Other devs won't know what PEA is.

caojoshua commented on src/hotspot/share/opto/partialEscape.hpp:
```diff
@@ -93,6 +95,21 @@ class VirtualState: public ObjectState {
 
+  bool equals(const VirtualState* rhs) {
+    if (this != rhs) {
+      if (_lockcnt != rhs->_lockcnt) return false;
```
NIT: prefer `&&` on the three `return false` blocks

caojoshua commented on src/hotspot/share/opto/partialEscape.hpp:
```diff
@@ -273,6 +294,10 @@ class PEAState {
   void add_new_allocation(GraphKit* kit, Node* obj);
 
   Node* materialize(GraphKit* kit, Node* var);
+  // passive_materialize case-a
```
We should not write case-a or case-b. I also do not think that `passive materialization` is the right term we should use in code. It is a made-up term.

I suggest `merge_incoming_virtual` and `merge_incoming_escaped`.

navyxliu replied:
good idea.

caojoshua commented on src/hotspot/share/opto/partialEscape.hpp:
```diff
@@ -317,22 +353,26 @@ class PEAState {
+  using ObjColorTable = ResourceHashtable<ObjID, bool, 23, /*table_size*/
+                                 AnyObj::RESOURCE_AREA, mtCompiler>;
+
+  void dfs(GraphKit* kit, ObjColorTable& color, ObjID key, VirtualState* vs, RegionNode* region, RegionNode* shim, const GrowableArray<uint>& paths);
```
dfs? This does not tell me what the function does. Maybe `materialize_object_graph`?

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -615,30 +612,43 @@ void PEAState::put_field(GraphKit* kit, ciField* field, Node* objx, Node* val) {
-  ObjID alloc = pea->is_alias(var);
+// clone AllocateNode
+Node* PEAState::clone_object(GraphKit* kit, ObjID alloc) {
+  auto pea = kit->PEA();
```
Not sure of OpenJDK's code style policy on `auto`, but feels unnecessary here.

caojoshua commented on src/hotspot/share/opto/partialEscape.hpp:
```diff
@@ -317,22 +353,26 @@ class PEAState {
-  void validate() const;
-#endif
+protected:
+  using ObjColorTable = ResourceHashtable<ObjID, bool, 23, /*table_size*/
```
comment for purpose of `ObjColorTable`? Maybe `ObjBitMap` makes more sense.

navyxliu replied:
using BitMap?

caojoshua commented on src/hotspot/share/opto/partialEscape.hpp:
```diff
@@ -273,6 +294,10 @@ class PEAState {
 
   Node* materialize(GraphKit* kit, Node* var);
+  // passive_materialize case-a
+  Node* passive_materialize(GraphKit* kit, Node* var, PEAState& newin, SafePointNode* newin_map, PhiNode* phi, RegionNode* r, uint pnum);
```
Has no callers? Is this PR actually for case-b? Your use a/b in code, but `case1` in the PR title.

navyxliu replied:
yes. it's for Case-b.

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,241 @@ Node* PEAState::get_java_oop(ObjID id) const {
+  return retval;
+}
+
+Node* PEAState::passive_materialize_for_predecessors(GraphKit* kit, Node* var, RegionNode* r, uint pnum) {
```
Need comment explaining the cases

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,241 @@ Node* PEAState::get_java_oop(ObjID id) const {
+      }
+    }
+  }
+  //void PEAState::dfs(GraphKit* kit, ObjColorTable& color, ObjID key, VirtualState* vs, RegionNode* region, RegionNode* shim, const GrowableArray<uint>& paths);
```
delete

navyxliu commented:
We got 3 new failures we have never seen before. 
https://github.com/navyxliu/jdk/actions/runs/6857581528

```
jdk/classfile/LvtTest
tools/javac/parser/JavacParserTest
java/text/Format/NumberFormat/NumberRoundTrip
```

Commit:
Remove shim region and add more comments.

Commit:
Replace ReosurceBitMap with Unique_Node_List.

caojoshua commented on src/hotspot/share/opto/cfgnode.cpp:
```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+      // | Region         |
+      // +----------------+
+      //
+      // we need to normalize the region node, and its phi nodes.
```
What does it mean to normalize a region node? What is `CFG node` and `cfg n`?

navyxliu replied:
let's say CFG Node = 100. CFG n = 101 and region is 200. 
because of passive materialization, we have a region node like this.  CFG Node 100 covers 3 paths of 200 Region.  those 3 paths share the same 'materialization code snippet'.
 
```
200 Region === [200, 100, 100, 100, 101]
```

we expect to see.
```
200 Region === [200, 100, 101]
```

This is known as 'normalization'.  200 Region === [200, 100, 100, 100, 101] isn't valid form for build_cfg. C2 codegen will run into problem. that's why we have to normalize in optimizer.

caojoshua replied:
OK makes sense. I think the labels are a little confusing. Can we just say we are removing duplicate inputs? I'm not sure 'normalization' is a standard term here, I don't see it in code.

navyxliu replied:
Not duplicated inputs. essentially, we hoist a common factor. 
Each input represents one predecessor of the basic block. we materialize the object once for all of them.

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,245 @@ Node* PEAState::get_java_oop(ObjID id) const {
+}
+// Passive Materialization
+// ------------------------
+// Materialize an object at the phi node because at least one of its predecessors has materialized the object.
```
> Materialize an object at the phi node

I don't think this is the correct way to describe it. We want to materialize in the shim region.

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,245 @@ Node* PEAState::get_java_oop(ObjID id) const {
+  GraphKit shim_kit(map->jvms());
+  bool unique = paths.length() <= 1;
+
+  // shim only covers selective paths, so we have to rewrite inputs if they are phi nodes.
```
What is a selective path?

navyxliu replied:
selective paths are all merged predecessors.  there are still empty inputs in the region.  they are unseen paths ,or yet-parsed predecessors.

Commit:
Fix 6 ctw errors.

Add an artifactual scope for ResourceMark. it's because ReplaceList in
materialize() also use GrowableBitMap. it's incorrect to release spaces
for them.

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,245 @@ Node* PEAState::get_java_oop(ObjID id) const {
+      map->set_req(i, new_mem);
+    } else if (n->is_Phi() && n->in(0) == r) {
+      if (unique) {
+        map->set_req(i, n->in(paths.at(0)));
```
Would it make sense to make this a phi with a single input? Because if you merge in a virtual object that needs to be passively materialized, you will use the same shim and will need to create a phi at that point.

caojoshua commented on src/hotspot/share/opto/cfgnode.cpp:
```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+          for (int j = paths.length()-1; j>=0; --j) {
+            nc.del_req(paths.at(j));
+          }
+          return this;
```
Is this an early return? We can only remove one set of dupllicates. We` can have multiple sets of duplicates. For example

```
200 Region === [200, 100, 100, 101, 101]
```

I don't think PEA will do this, but maybe it would make sense for common RegionNode code.

navyxliu replied:
this is done by PhaseGVN. transform() keep calling ideal() until it turns nullptr. 
```cpp
  Node* k = n;
  Node* i = apply_ideal(k, /*can_reshape=*/false);
  NOT_PRODUCT(uint loop_count = 1;)
  while (i != nullptr) {
    assert(i->_idx >= k->_idx, "Idealize should return new nodes, use Identity to return old nodes" );
    k = i;
    i = apply_ideal(k, /*can_reshape=*/false);
    NOT_PRODUCT(loop_count++;)
  }
```

caojoshua commented on src/hotspot/share/opto/cfgnode.cpp:
```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
+          for (DUIterator_Fast kmax, k = fast_outs(kmax); k < kmax; k++) {
+            Node* phi = fast_out(k);
+
+            if (phi != nullptr && phi->is_Phi() && phi->req() == req()) {
```
Is `phi->req() == req()` always true? Phi's should have the same number of inputs as the region node.

caojoshua commented on src/hotspot/share/opto/parseHelper.cpp:
```diff
@@ -869,9 +865,248 @@ Node* PEAState::get_java_oop(ObjID id) const {
+}
 
-void AllocationStateMerger::merge(PEAState& newin, GraphKit* kit, RegionNode* region, int pnum) {
+// update the internal phi node in virtual objects.
```
Can we have a better name for `color`. I have no idea what it means.

Can we describe what needs to be updated?

navyxliu replied:
for DFS, we need to prevent from revisiting a node which has been visited. 
the standard DFS algorithm using coloring scheme: white (unvisited), gray(processing) and black (processed). 

this is simplified version. we don't need to distinct gray and black. we just mark it visited.

Commit:
Merge branch 'PEA_beta' into PEA_shim2

Commit:
Don't lock down phi nodes of a MergeMemNode until we complete PEA
allocation state merge.

Commit:
Do passive materialization at AllocationStateMerger.

Commit:
Split AllocationStateMerger::merge() into 3 phases.

1. only handle passive materialization
2. merge Escaped objects by creating phi nodes.
3. merge virtual objects.

The reason we have to do this into phases because objects may hold
reference to others. if so, order matters. we can't sort them beforehand
because object graph may have cycle.

Commit:
Copy allocation State from ex_map in throw_to_exit().

This patch fixed compiler/arraycopy/TestCloneWithStressReflectiveCode.java

Commit:
Merge branch 'PEA_beta' into PEA_shim2

Commit:
Fix release build.

Commit:
Split merge virtual state into 2 passes.

It's because it's possible that there are identical phi nodes for
different fields. if we do global value numbering field by field, we
can't process the later field.

Commit:
Don't update local symbol table in merge_common.

It's because we intercept the phi creation aftr Parse::ensure_phi.
ensure_phi checks the type of map->in(i) from ciTypeFlow and don't
create phi if it's invalid.

Commit:
JVM-2544: java/util/stream/test/org/openjdk/tests/java/util/stream/FlatMapOpTest

Commit:
JVM-2545: bad graph detetect in java/security/MessageDigest/TestDigestIOStream

Commit:
Merge branch 'PEA_beta' into PEA_shim2

Commit:
Merge branch 'PEA_beta' into PEA_shim2

Commit:
Merge branch 'PEA_beta' into PEA_shim2
